branchsearch.cpp

```diff
@@ -93,6 +93,42 @@ void filter_branches(vector<string> &all_branches, char* regex_value, vector<str
     }
 }
 
+bool best_match(vector<string> &all_branches, char* pattern, string& res){
+    // pattern to lowercase
+    transform(pattern, pattern+strlen(pattern), pattern, ::tolower);
+
+    auto better_match = [pattern]( const string& a, const string& b) {
+        // the one starting with the pattern is prefered
+        if(a.find(pattern) == 0 and b.find(pattern) != 0) return true;
+        if(b.find(pattern) == 0 and a.find(pattern) != 0) return false;
+
+        // the one closer in length to the pattern is prefered
+        if(a.size()-strlen(pattern) < b.size()-strlen(pattern)) return true;
+        if(b.size()-strlen(pattern) < a.size()-strlen(pattern)) return false;
+
+        // tiebreaker
+        return a < b;
+    };
+
+    bool found = false;
+    int index = -1;
+    for (int i = 0; i< all_branches.size(); i++){
+        string b = all_branches[i];
+        // branch to lowercase
+        transform(b.begin(), b.end(), b.begin(), ::tolower);
+        if (b.find(pattern) != string::npos) {
+            if(!found or better_match(b, res)){
+                found = true;
+                index = i;
+                res = b;
+            }
+        }
+    }
+
+    if(found) res = all_branches[index];
```

Comment:
Why are you doing this? I mean, res already has the best you have found, right? Also, seems like having `index` _and_ `found` is unnecessary. Having an `index >= 0` is sign you have found something you can use.

Reply:
> Why are you doing this? I mean, res already has the best you have found, right?

It was transformed to lower case. I need to get the original branch name

Reply:
You are right.